src/PluginInterface.ts

```diff
@@ -0,0 +1,44 @@
+import { CompilerPlugin } from './interfaces';
+import { Logger } from './Logger';
+
+// inspiration: https://github.com/andywer/typed-emitter/blob/master/index.d.ts
+
+export type Arguments<T> = [T] extends [(...args: infer U) => any]
+    ? U
+    : [T] extends [void] ? [] : [T];
+
+export default class PluginInterface<T extends CompilerPlugin = CompilerPlugin> {
```

Comment:
How about we call this `Plugin`, since it's actually a class instance and _not_ an interface.

Reply:
Not necessarily the best name but ot's not the plugin, it's the interface (not in an OOP way) to the plugins that the compiler uses to emit events.